src/OpenTelemetry.Api/Logs/LogRecordAttributeList.cs

```diff
@@ -207,12 +208,12 @@ public void RecordException(Exception exception)
     /// <inheritdoc/>
     readonly IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
 
-    internal readonly List<KeyValuePair<string, object?>>? Export(ref List<KeyValuePair<string, object?>>? attributeStorage)
+    internal readonly IReadOnlyList<KeyValuePair<string, object?>> Export(ref List<KeyValuePair<string, object?>>? attributeStorage)
     {
         int count = this.count;
         if (count <= 0)
         {
-            return null;
+            return Empty;
```

Comment:
How does this work with `LoggerSdk.EmitLog`? This would be changing the behavior of `LoggerSdk.EmitLog` to  return an empty list instead of `null` when the attribute count is zero. Is that what we want? https://github.com/open-telemetry/opentelemetry-dotnet/blob/303a35789508a4f7ad3ef8dd968e68fa62eef109/src/OpenTelemetry/Logs/LoggerSdk.cs#L56

`LogRecord.Attributes` description says that its value is set depending upon the options: `IncludeAttributes` or `ParseStateValues`. Would these options be considered when using `LoggerSdk.EmitLog`?

Reply:
There's a lot to unpack here 🤣 

> How does this work with LoggerSdk.EmitLog? This would be changing the behavior of LoggerSdk.EmitLog to return an empty list instead of null when the attribute count is zero. Is that what we want?

I don't think it is an issue. It is all net-new so up to us to define the behavior. When calling `Logger.EmitLog` (at the moment) there are always attributes. It is either user-supplied (may have 0 count or more) or it is `default` (0 count). It is a struct so there isn't a way to pass `null`. With this change the `LogRecord` will mirror that behavior. It will always have attributes either empty (0) or some list using the pool storage.

I made this change because the exporter which causes the 1.5.1 breakfix is doing...

`var list = (IReadOnlyList<KeyValuePair<string, object>>)logRecord.State;`

It isn't checking for `null` and it isn't looking at `Attributes` or `StateValues`. If some users take a newer SDK and starts using some `Logger.EmitLog` thing that exporter would break again. I needed to make sure this path also had a valid `logRecord.State`. Since `State` would have a value I decided `Attributes` / `StateValues` should as well. It felt confusing the other way around.

> LogRecord.Attributes description says that its value is set depending upon the options: IncludeAttributes or ParseStateValues. Would these options be considered when using LoggerSdk.EmitLog?

`OpenTelemetryLoggerOptions` only applies to the `ILogger` integration. I'll open a follow-up PR to clarify the XML comments on `LogRecord` because you are right they are kind of misleading.

Now you can use `LogRecordAttributeList` with the `ILogger` integration! You can do `logger.Log<LogRecordAttributeList>(...)` which is using `LogRecordAttributeList` as `TState` in the `ILogger` API. This is really the only way at the moment to get truly allocation-free logging in OpenTelemetry .NET. If you set `OpenTelemetryLoggerOptions.IncludeAttributes = false` AND you use `LogRecordAttributeList` as the `TState` THEN `IncludeAttributes` is respected and the `LogRecord` ends up with `null` values for `State`, `Attributes`, and `StateValues`.